src/vector/transaction.rs

```diff
@@ -177,11 +177,31 @@ impl Dataset {
         }
         Ok(Transaction::new(self))
     }
+
+    /// Optionally start a transaction before running `func` for performance
+    ///
+    /// This uses transaction if the dataset supports it, otherwise it
+    /// runs the `func` function as it is on the dataset.
+    pub fn maybe_transaction(&mut self, func: impl Fn(&Dataset) -> Result<()>) -> Result<()> {
+        let force = 0; // since this is for speed
+        let rv = unsafe { gdal_sys::GDALDatasetStartTransaction(self.c_dataset(), force) };
+        let res = func(self);
+        if rv == OGRErr::OGRERR_NONE {
+            let rv = unsafe { gdal_sys::GDALDatasetCommitTransaction(self.c_dataset()) };
+            if rv != OGRErr::OGRERR_NONE {
+                return Err(GdalError::OgrError {
+                    err: rv,
+                    method_name: "GDALDatasetCommitTransaction",
+                });
+            }
+        }
```

Comment:
I don't think the implementation is correct yet.

Something like this is required to handle at least basic errors.

```suggestion
        let force = 0; // since this is for speed
        let rv = unsafe { gdal_sys::GDALDatasetStartTransaction(self.c_dataset(), force) };
        let res = func(self);
        if rv == OGRErr::OGRERR_NONE {
            // check for the result of the write process of the user,
            // if it's an error we want to rollback, otherwise commit
            match &res {
                Ok(_) => {
                    let rv = unsafe { gdal_sys::GDALDatasetCommitTransaction(self.c_dataset()) };
                    if rv != OGRErr::OGRERR_NONE {
                        // if commit returns an error, at least try to rollback
                        // to keep the dataset in a usable state
                        let rv2 = unsafe { gdal_sys::GDALDatasetRollbackTransaction(self.c_dataset()) };
                        if rv2 != OGRErr::OGRERR_NONE {
                            // if rollback also returns an error we
                            // want to return the rollback error
                            return Err(GdalError::OgrError {
                                err: rv2,
                                method_name: "GDALDatasetRollbackTransaction",
                            });
                        } else {
                              // otherwise return the commit error
                              return Err(GdalError::OgrError {
                                  err: rv,
                                  method_name: "GDALDatasetCommitTransaction",
                              });
                        }
                }
                // if the user returned an error, rollback the changes
                Err(_) => {
                    let rv = unsafe { gdal_sys::GDALDatasetRollbackTransaction(self.c_dataset()) };
                    if rv != OGRErr::OGRERR_NONE {
                        return Err(GdalError::OgrError {
                            err: rv,
                            method_name: "GDALDatasetRollbackTransaction",
                    });
                }
            }
            res
        }
```